README.md
# Traffic Prediction xApp



## A. Directory Structure
```bash
Traffic-Prediction-xApp
|__ src # The folder of source code of xApp do
|   |__ cleansing.py # Make the data more trainable by cleansing it in scaler and create the sequences for training
|   |__ database.py # Connecting to the database (InfluxDB) 
|   |__ exceptions.py # Exception When The Code Not Execution Properly
|   |__ main.py # Main File To Execution the xApp
|   |__ model_load.py # Load the model that already exist
|   |__ training.py # Train the model if the model isn't already exist
|__ xapp-descriptor # The folder of description of the xApp in Kubernetes
|   |__ config.json # Configuration file for description of xApp
|   |__ schema.json # Control section of container in one node of kubernetes
|__ container-tag.yaml # Version Container
|__ Dockerfile # DockerFile for running the xApp in container image of xApp
|__ INFO.yaml # Information of the code
|__ LICENSE # License of the code
|__ local.rt # Set the message type that will be used as RMR seed
|__ rmr-version.yaml # Set the RMR version
|__ setup.py # Setup the xApp as Python Image and set library that necessary
|__ tox.ini # Description of xApp Package
```



## B. Requirements
### B.1 Python Library
- Python 3.x
- NumPy
- Pandas
- TensorFlow
- RICxAPPFrame
- JobLib
- MDCLogpy
- InfluxDB
- Scikit-Learn
### B.2 Image and Container
- KUBEV v1.16.0
- HELMV 3.5.4
- DOCKERV 20.10.21


## C. Running The xApp and Deploy it into local K8S
### Running in Docker
```bash
docker build -t tp-xapp:latest -f  Dockerfile .
docker run -i --net=host tp-xapp:latest
```
### xApp Deployment
```bash
docker build -t tp-xapp:latest -f  Dockerfile .
export CHART_REPO_URL=http://0.0.0.0:8090
dms_cli onboard ./xapp-descriptor/config.json ./xapp-descriptor/schema.json
dms_cli install tp-xapp 1.0.0 ricxapp
```


## D. Data Cleansing
Not all the data can be directly trained in any model of intelligence control. They should be cleaned first before utilize as reference data.
```python
class Cleansing(object):
    def __init__(self, df = None, lookback = 10):
      self.df = df
      self.lookback = lookback
```
Parameters:
1. `df` (DataFrame, required): The Data that performs the train.
2. `lookback` (Int, required) : The time step before of the train.
### Example Usage
```python
cleaner = Cleansing(df,10)
cleaner.normalizer()
input_train, output_train = cleaner.create_sequences()
```



## E. Model Load and Training
Load model and training are two different classes. The load model Class loads an existing model, but Training perfroms training to generate a model
### E.1 Model Load
Besides of load the model, This class also predict the data from the own methods
```python
class ModelLoad(object):
    r""" Load the model

    Parameters
    ----------
    tfLite:bool
    """

    def __init__(self, tfLite = True):
        self.tfLite = tfLite
        self.load_model()
        self.load_scale()
```
Parameters:
1. `tfLite` (bool, required): The option where you stored the model, the lite or the original, True if using the lite model.

Other parameters:
2. `inputs` (DataFrame, required): The parameters use in predict the data using methods predits(DataFrame)
#### Example Usage
```python
md = ModelLoad(tfLite=False)
prediction = md.predict(DataFrame)
```
### E.2 Model Training
```python
class Training(object):
    r""" Training the model

    Parameters
    ----------
    df:DataFrame
    """
    def __init__(self,df):
        self.df = df
```
Parameters:
1. `df` (DataFrame, required): The Data that performs the train.
#### Models 3D-CNN LSTM
3D-CNN LSTM (Convolutional Neural Network Long Short-Term Memory) is a hybrid neural network architecture that combines the strengths of 3D Convolutional Neural Networks (3D-CNNs) and Long Short-Term Memory (LSTM) networks. This architecture is specifically designed for processing spatio-temporal data, such as video sequences or 3D time series data.
```python
model_3dcnn = tf.keras.Sequential([
    tf.keras.layers.Conv3D(filters=32, kernel_size=(3, 3, 3), activation='relu', input_shape=(time_steps, 7, 1, 1), padding='same'),
    tf.keras.layers.MaxPooling3D(pool_size=(2, 2, 2), strides=(1, 1, 1), padding='same'),
    tf.keras.layers.Conv3D(filters=64, kernel_size=(3, 3, 3), activation='relu', padding='same'),
    tf.keras.layers.MaxPooling3D(pool_size=(2, 2, 2), strides=(1, 1, 1), padding='same'),
    tf.keras.layers.Flatten()
])
lstm_input_shape = model_3dcnn.output_shape[1:]
lstm_model = tf.keras.Sequential([
    tf.keras.layers.Reshape((time_steps, -1)),
    tf.keras.layers.LSTM(256, activation='relu', input_shape=lstm_input_shape, return_sequences = True),
    tf.keras.layers.LSTM(128, activation='relu'),
    tf.keras.layers.Dense(64, activation='relu'),
    tf.keras.layers.Dense(time_steps * 7, activation='linear'),
    tf.keras.layers.Reshape((time_steps, 7))
])
combined_input = tf.keras.layers.Input(shape=(time_steps, 7, 1, 1))
cnn_output = model_3dcnn(combined_input)
lstm_output = lstm_model(cnn_output)
combined_model = tf.keras.models.Model(inputs=combined_input, outputs=lstm_output)
combined_model.compile(optimizer=optimizer, loss=loss, metrics=['accuracy'])
```
The model utilize the 3D-CNN-LSTM because the data itself that from many base station and gather together in one data. The base station represent each matrix that could be visualize as big matrix in one data. Thus, to make matrix represents each base, it converts to 3D matrix. The best methods to predict is using the 3D-CNN. Afterthat, LSTM helps to predict future steps ahed of what base station output next.


## F. Database (InfluxDB)
Data that stored in database from UE fetch via InfluxDB.
```python
class Database(object):
    r"""
    Fetch the data from InfluxDB. InfluxDB has method get_field, get_measurement, get_start, get_stop, get_time, and get_value

    Parameters
    ----------
    url: String
    token: String
    org: String
    bucket: String
    """
    def __init__(self, url='', token='', org='', bucket='kpimon'):
        self.url = url
        self.token = token
        self.org = org
        self.bucket = bucket
        self.client = None
        self.config()
```
Parameters:
1. `url` (String, required) : The url of database
2. `token` (String, required) : The token of influxDB
3. `org` (String, required) : The org that want to fetch
4. `bucket` (String, required) : The bucket that want to fetch



docs/release-notes.rst
.. This work is licensed under a Creative Commons Attribution 4.0 International License.
.. SPDX-License-Identifier: CC-BY-4.0
.. Copyright (c) 2020 HCL Technologies Limited.
.. Copyright (C) 2020 AT&T Intellectual Property

Release Notes
===============

All notable changes to this project will be documented in this file.

The format is based on `Keep a Changelog <http://keepachangelog.com/>`__
and this project adheres to `Semantic Versioning <http://semver.org/>`__.

[0.0.1] - 2023-08-02
--------------------
* Initial mock version (`RICAPP-107 <https://jira.o-ran-sc.org/browse/RICAPP-107>`_)


src/main.py
import json
import os
import time
import pandas as pd
import schedule
from mdclogpy import Logger
from training import Training
from database import Database
from model_load import ModelLoad
from ricxappframe.xapp_sdl import SDLWrapper
from ricxappframe.xapp_frame import Xapp

db = None
df = None
sdl = SDLWrapper(use_fake_sdl=True)

logger = Logger(name=__name__)


def entry(self):
    """The entry file of running the xApp. This is the xApp general function will run. This function run every 1 second to predict and send it."""
    
    connectdb()
    train_model()
    load_model()
    schedule.every(1).seconds.do(predict, self)
    while True:
        schedule.run_pending()


def load_model():
    """The model will be load if there model in ./src"""
    
    global md
    logger.info("Load the Model")
    md = ModelLoad(tfLite=False)

def train_model():
    """The model that will be trained while the model doesn't exist in ./src file"""
    
    if not os.path.exists(os.getcwd()+'/src/model.h5') and not os.path.exists(os.getcwd()+'/src/model.tflite'):
        logger.info("Start Training The Model")
        df = db.queries("|> range(start: -5h)")
        if len(df) == 0:
            logger.info("No Data in Database")
        else:
            mt = Training(df)
            mt.train()
            mt.model_saved()

def predict(self):
    """Prediction function from fetching last 1 second data and decode the prediction into JSON for sending it to network slicing xApp"""
    
    column_names = ['DRB_UEThpUl', 'Viavi_Nb1_Rsrp', 'Viavi_Nb1_Rsrq', 'Viavi_Nb2_Rsrp', 'Viavi_Nb2_Rsrq', 'Viavi_UE_Rsrp', 'Viavi_UE_Rsrq']
    df = db.queries("|> range(start: -1s)")
    val = None
    if len(df) > 0 and df is not None:
        val = md.predict(df)
        val = val[0]
        result_send = dict()
        for column in range(len(column_names)):
            result_send[column_names[column]] = val[:,column]
        result = json.loads(result_send.to_json(orient='records'))
        val = json.dumps(result).encode()
    else:
        logger.warning("No data in last 1 second")
        time.sleep(1)
    if (val is not None) and (len(val) > 0):
        msg_to_xapp(self, val)

def msg_to_xapp(self, val):
    """Sending the database to Slicing Network xApp as 80911 message type in 30 data prediction ahead"""
    
    logger.debug("Sending to Slicing Network xApp")
    success = self.rmr_send(val, 80901)
    if success:
        logger.info("Message to Slicing Network xApp: message sent Successfully")
    for (summary, sbuf) in self.rmr_get_messages():
        logger.info("Received acknowldgement from SLICING NETWORK xAPP: {}".format(summary))
        self.rmr_free(sbuf)


def connectdb(thread=False):
    """Connection to the database"""
    
    global db, df
    db = Database()
    success = False
    while not success:
        success = db.connect()


def start(thread=False):
    """Starting of xApp. xApp will execute this function. This Function initiates xApp API and runs the entry() using xapp.run()"""
    
    xapp = Xapp(entrypoint=entry, rmr_port=4560, use_fake_sdl=False)
    xapp.logger.debug("Traffic Prediction xApp starting")
    xapp.run()


src/database.py
#!/usr/bin/python3

import time
from mdclogpy import Logger
from influxdb_client import InfluxDBClient
from configparser import ConfigParser
from influxdb_client.rest import ApiException
import pandas as pd

logger = Logger(name=__name__)

class Database(object):
    """Fetch the data from InfluxDB. InfluxDB has method get_field, get_measurement, get_start, get_stop, get_time, and get_value
        
        Args:
            url (String): The url of the influxDB that you want to link it.
            token (String): The tokenize of influxDB.
            org (String): The destination org that you want to fetch
            bucket (String): The destination bucket that you want to fetch
    """
    def __init__(self, url='', token='', org='', bucket=''):
        """Initialize the Database instance.
        
        Args:
            url (String): The url of the influxDB that you want to link it.
            token (String): The tokenize of influxDB.
            org (String): The destination org that you want to fetch
            bucket (String): The destination bucket that you want to fetch
        """
        
        self.url = url
        self.token = token
        self.org = org
        self.bucket = bucket
        self.client = None
        self.config()

    def connect(self):
        """Connect to the link database that has been declare"""
        
        self.config()
        if self.client is not None:
            self.client.close()
        try:
            self.client = InfluxDBClient(url=self.url, token=self.token, org=self.org)
            logger.info("Connected to Influx Database")
            return True
        except ApiException as e:
            logger.error(f"Error connecting to InfluxDB server: {e}")
            time.sleep(30)
        except Exception as e:
            logger.error(f"An error occurred: {e}")
            time.sleep(30)
    
    def disconnect(self):
        """Disconnect from the database"""
        
        if self.client is not None:
            self.client.close()

    def queries(self,params):
        """Initialize the Database instance.
        
        Args:
            params (String): The argument of parameters for fetch last data,.e.g. [|> range(start: -1m)]
        """

        try:
            column_names = ['DRB_UEThpUl', 'Viavi_Nb1_Rsrp', 'Viavi_Nb1_Rsrq', 'Viavi_Nb2_Rsrp', 'Viavi_Nb2_Rsrq', 'Viavi_UE_Rsrp', 'Viavi_UE_Rsrq']
            value_result = dict()
            for column in column_names:
                value_result[column] = []
            queriese = f'from(bucket: {self.bucket}) {params}'
            tables = self.client.query_api().query(queriese, org=self.org)
            if len(tables) > 0:
                for table in tables:
                    for record in table.records:
                        if record.get_measurement() == 'UeMetrics' and record.get_field() in column_names:
                            value_result[record.get_field()].append(record.get_value())
            return pd.DataFrame(value_result)
        except Exception as e:
            logger.error(f"An error occurred: {e}")
            return None

    def config(self):
        """Declare the data hasbeen store from .ini file and restore it in this function such as url,token,org, and bucket"""
        
        cfg = ConfigParser()
        cfg.read('./src/tf_config.ini')
        for section in cfg.sections():
            if section == 'InfluxDB':
                self.url = cfg.get(section, "url")
                self.token = cfg.get(section, "token")
                self.org = cfg.get(section, "org")
                self.bucket = cfg.get(section, "bucket")

# if __name__ == "__main__":
#     db = Database()
#     db.connect()
#     res = db.queries(params="|> range(start: -0.3s)")
#     db.disconnect()


src/cleansing.py
from sklearn.preprocessing import MinMaxScaler
import pandas as pd
import numpy as np
import joblib

class Cleansing(object):
    """Clean the data from DataFrame data type (Pandas Python Library). Besides
        It also normalize the data and separate it from sequence that you want.
        
        Args:
            df (DataFrame): An datasets that created by DataFrame by python library
            lookback (int): An lookback data will separated by lookback sequences
    """
    def __init__(self, df = None, lookback = 30):
        """Initialize the Cleansing instance.
        
        Args:
            df (DataFrame): An datasets that created by DataFrame by python library
            lookback (int): An lookback data will separated by lookback sequences
        """
        self.df = df
        self.lookback = lookback

    def change_to_field(self):
        """Change the data into the DataFrame that training want. It uses when use the offline data."""
        
        unique_field = self.df['_field'].unique()
        field_idb = self.df.groupby('_field')
        dict_new_analytics = dict()
        for i in range(len(unique_field)):
           dict_new_analytics[unique_field[i]] = list(field_idb.get_group(unique_field[i])["_value"])
        df_new_analytics = pd.DataFrame(dict_new_analytics)
        df_new_analytics = df_new_analytics.drop(['DRB_UECqiDl','DRB_UECqiUl','DRB_UEThpDl','QosFlow_TotPdcpPduVolumeDl','RRU_PrbUsedDl','RRU_PrbUsedUl','TB_TotNbrDl','TB_TotNbrUl','Viavi_Geo_x','Viavi_Geo_y','Viavi_Geo_z','Viavi_Nb1_RsSinr','Viavi_Nb2_RsSinr','Viavi_QoS_5qi','Viavi_QoS_Gfbr','Viavi_QoS_Mfbr','Viavi_QoS_Priority','Viavi_QoS_Score','Viavi_QoS_TargetTput','Viavi_UE_anomalies','Viavi_UE_targetThroughputDl','Viavi_UE_targetThroughputUl'],axis=1)
        df_new_analytics['DRB_UEThpUl'] = df_new_analytics['DRB_UEThpUl'].apply(lambda x: x*1024)
        return df_new_analytics

    def normalizer(self):
        """Normalize the data into scale 0 to 1. The transform data are stored in /src directory. Use ./src/scale to restore the scaler"""
        
        column_df_analytics = list(self.df.columns)
        scaler = MinMaxScaler(feature_range=(0, 1))
        scaler_saved = scaler.fit(self.df[column_df_analytics])
        joblib.dump(scaler_saved, './src/scale')
        self.df[column_df_analytics] = scaler.transform(self.df[column_df_analytics])

    def create_sequences(self):
        """Create the data into desire data need in the models (None,7,1,1) also the output (None,7) sequences"""
        
        samples = len(self.df.values) - self.lookback + 1
        input_data = np.zeros((samples, self.lookback, 7, 1, 1))
        output_data = np.zeros((samples, self.lookback, 7))
        for i in range(samples):
            input_data[i] = self.df.values[i:i+self.lookback].reshape((self.lookback, 7, 1, 1))
            output_data[i] = self.df.values[i:i+self.lookback].reshape((self.lookback, 7))
        return input_data,output_data


src/model_load.py
import joblib
from mdclogpy import Logger
import tensorflow as tf

logger = Logger(name=__name__)


class ModelLoad(object):
    """Load the model if exist in ./src file directory
        
        Args:
            tfLite (bool): A choose of load data depends on what the model saved in the .tf or .tflite. True if save in .tflite and otherwise.
    """

    def __init__(self, tfLite = True):
        """Initialize the Cleansing instance.
        
        Args:
            tfLite (bool): A choose of load data depends on what the model saved in the .tf or .tflite. True if save in .tflite and otherwise.
        """
        self.tfLite = tfLite
        self.load_model()
        self.load_scale()
    
    def load_model(self):
        """Load the model either tfLite or tf depend on what model saved in ./src file directory"""
        
        try:
            if self.tfLite:
                self.tfLite = True
                self.model = tf.lite.Interpreter(model_path='./src/model.tflite')
                self.model.allocate_tensors()
            else:
                self.tfLite = False
                self.model = tf.keras.models.load_model('./src/model.h5')
        except FileNotFoundError:
            logger.error("Model Does not exsist")

    def load_scale(self):
        """Load the scale that has been store by normalized function"""
        
        try:
            with open('./src/scale', 'rb') as f:
                self.scale = joblib.load(f)
        except FileNotFoundError:
            logger.error("Scale file does not exsist")

    def predict(self, inputs):
        """Prediction function from dataframe arguments.

        Args:
            inputs (DataFrame): A data represent to predict in DataFrame from Pandas Python Library.
        """
        
        pred = None
        if self.tfLite:
            input_details = self.model.get_input_details()
            output_details = self.model.get_output_details()
            self.model.set_tensor(input_details[0]['index'], inputs)
            self.model.invoke()
            pred = self.model.get_tensor(output_details[0]['index'])
        else:
            pred = self.model.predict(inputs)
        pred = self.scale.inverse_transform(pred)
        return pred


src/training.py
import tensorflow as tf
from database import Database
from cleansing import Cleansing

class Training(object):
    """Training the data if the model doesn't exist
        
        Args:
            df (DataFrame): An datasets that want to be train to outcome the model
    """
    
    def __init__(self,df):
        """Initialize the Cleansing instance.
        
        Args:
            df (DataFrame): An datasets that want to be train to outcome the model
        """
        self.df = df

    def create_model_forecasting(self,time_steps=30,optimizer='adam',loss='mean_squared_error'):
        """Create the model summary with 3D-CNN-LSTM
        
        Args:
            time_steps (int): A lookback training from time_step before and predict time_step ahead.
            optimizers (String): An optimizer of the model. It effects to how fast the model is created (For Further Information what kind of opimizer, you can see the tensorflow websites)
            loss (String): A loss function. as machine learning or deep learning are fixed the model from loss, the loss function also determine how fast the model is created (For Further Information what kind of loss function, you can see the tensorflow websites)
        """
        model_3dcnn = tf.keras.Sequential([
            tf.keras.layers.Conv3D(filters=32, kernel_size=(3, 3, 3), activation='relu', input_shape=(time_steps, 7, 1, 1), padding='same'),
            tf.keras.layers.MaxPooling3D(pool_size=(2, 2, 2), strides=(1, 1, 1), padding='same'),
            tf.keras.layers.Conv3D(filters=64, kernel_size=(3, 3, 3), activation='relu', padding='same'),
            tf.keras.layers.MaxPooling3D(pool_size=(2, 2, 2), strides=(1, 1, 1), padding='same'),
            tf.keras.layers.Flatten()
        ])
        lstm_input_shape = model_3dcnn.output_shape[1:]
        lstm_model = tf.keras.Sequential([
            tf.keras.layers.Reshape((time_steps, -1)),
            tf.keras.layers.LSTM(256, activation='relu', input_shape=lstm_input_shape, return_sequences = True),
            tf.keras.layers.LSTM(128, activation='relu'),
            tf.keras.layers.Dense(64, activation='relu'),
            tf.keras.layers.Dense(time_steps * 7, activation='linear'),
            tf.keras.layers.Reshape((time_steps, 7))
        ])
        combined_input = tf.keras.layers.Input(shape=(time_steps, 7, 1, 1))
        cnn_output = model_3dcnn(combined_input)
        lstm_output = lstm_model(cnn_output)
        self.combined_model = tf.keras.models.Model(inputs=combined_input, outputs=lstm_output)
        self.combined_model.compile(optimizer=optimizer, loss=loss, metrics=['accuracy'])

    def train(self,epochs=50,batch_size=32,val_split=0.2):
        """Train of the model that has been created from create_model_forecasting()
        
        Args:
            epochs (int): How many epochs train, it effets to how fast data will produce
            batch_size (int): A number of samples that will be used in a single iteration (forward and backward pass) of training in a machine learning algorithm, particularly in deep learning models like neural networks.
            val_split (float): A validation split in the context of training a machine learning or deep learning model. During the training process, it's common to split your available labeled data into three main subsets: the training set, the validation set, and the test set.
        """
        
        cleaner = Cleansing(self.df,10)
        cleaner.normalizer()
        input_train, output_train = cleaner.create_sequences()
        self.create_model_forecasting()
        self.combined_model.fit(input_train, output_train, epochs=epochs, batch_size=batch_size, validation_split=val_split)
    
    def model_saved(self):
        """Saved the model into .h5 (.tf) file"""
        
        self.combined_model.save(f"./src/model.h5")
    
    def model_saved_lite(self):
        """Saved the model into .tflite file"""
        
        converter = tf.lite.TFLiteConverter.from_keras_model(self.combined_model)
        tflite_model = converter.convert()
        with open('./src/model.tflite', 'wb') as f:
            f.write(tflite_model)
